tests/test_paradiag.py

```diff
@@ -309,6 +309,88 @@ def form_mass(u, h, v, q):
         assert (abs(uerr) < utol)
 
 
+@pytest.mark.parallel(nprocs=4)
+def test_Nitsche_BCs():
+    # test the linear equation u_t - Delta u = 0, with u_ex = exp(0.5*x + y + 1.25*t) and weakly imposing Dirichlet BCs
+    nspatial_domains = 2
+    degree = 1
+    nx = 10
+    dx = 1/nx
+    dt = dx
+    ensemble = fd.Ensemble(fd.COMM_WORLD, nspatial_domains)
+    mesh = fd.UnitSquareMesh(nx, nx, quadrilateral=False, comm=ensemble.comm)
+
+    x, y = fd.SpatialCoordinate(mesh)
+    n = fd.FacetNormal(mesh)
+    V = fd.FunctionSpace(mesh, "CG", degree)
+
+    w0 = fd.Function(V)
+    w0.interpolate(fd.exp(0.5*x + y))
+
+    def form_mass(q, phi):
+        return phi*q*fd.dx
+
+    def form_function(q, phi, t):
+        return fd.inner(fd.grad(q), fd.grad(phi))*fd.dx - fd.inner(phi, fd.inner(fd.grad(q), n))*fd.ds - fd.inner(q-fd.exp(0.5*x + y + 1.25*t), fd.inner(fd.grad(phi), n))*fd.ds + 20*nx*fd.inner(q-fd.exp(0.5*x + y + 1.25*t), phi)*fd.ds
+
+    # Parameters for the diag
+    sparameters = {
+        'ksp_type': 'preonly',
+        'pc_type': 'lu',
+        'pc_factor_mat_solver_type': 'mumps'}
+
+    solver_parameters_diag = {
+        "snes_linesearch_type": "basic",
+        'snes_atol': 1e-8,
+        # 'snes_monitor': None,
+        # 'snes_converged_reason': None,
+        'ksp_rtol': 1e-8,
+        # 'ksp_monitor': None,
+        # 'ksp_converged_reason': None,
+        'mat_type': 'matfree',
+        'ksp_type': 'gmres',
+        'pc_type': 'python',
+        'pc_python_type': 'asQ.DiagFFTPC',
+    }
+
+    M = [2, 2]
+    solver_parameters_diag["diagfft_block_"] = sparameters
+
+    alpha = 1.0e-3
+    theta = 0.5
+
+    PD = asQ.paradiag(ensemble=ensemble,
+                      form_function=form_function,
+                      form_mass=form_mass, w0=w0,
+                      dt=dt, theta=theta,
+                      alpha=alpha,
+                      time_partition=M, bcs=[], solver_parameters=solver_parameters_diag,
+                      circ=None, tol=1.0e-6, maxits=None,
+                      ctx={}, block_mat_type="aij")
+    PD.solve()
+
+    # check against initial conditions
+    aaos = PD.aaos
+    q_exact = fd.Function(V)
+    qp = fd.Function(V)
+    errors = asQ.SharedArray(M, comm=ensemble.ensemble_comm)
+    times = asQ.SharedArray(M, comm=ensemble.ensemble_comm)
+    for step in range(aaos.ntimesteps):
```

Comment:
I think you can just loop over the local timesteps here rather than over the global timesteps